aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,394 @@
+            if 1 != EVP_PKEY_get_raw_public_key(*self.context, null_mut(), &mut pubkey_len) {
+                pubkey_len.zeroize();
+                return Err(Unspecified);
+            }
+
+            let mut pubkey_bytes = vec![0u8; pubkey_len];
+            let pubkey_ctx_copy;
+
+            if 1 != EVP_PKEY_get_raw_public_key(
+                *self.context,
+                pubkey_bytes.as_mut_ptr(),
+                &mut pubkey_len,
+            ) {
+                return Err(Unspecified);
+            }
+            pubkey_ctx_copy = LcPtr::new(EVP_PKEY_kem_new_raw_public_key(
+                self.algorithm.nid(),
+                pubkey_bytes.as_mut_ptr(),
+                pubkey_len,
+            ))?;
+            Ok(KemPublicKey {
+                algorithm: self.algorithm.clone(),
+                context: pubkey_ctx_copy,
+                pub_key: pubkey_bytes.into(),
+            })
+        }
+    }
+
+    /// Performs the decapsulate operation using the current KEM private key on the given ciphertext.
+    ///
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM public key
+    /// associated with the current KEM private key.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn decapsulate<F, R>(&self, ciphertext: &mut [u8], kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8]) -> Result<R, Unspecified>,
+    {
+        unsafe {
+            let ctx = LcPtr::new(EVP_PKEY_CTX_new(*self.context, null_mut()))?;
+            let mut shared_secret_len: usize = 0;
+
+            if 1 != EVP_PKEY_decapsulate(
+                *ctx,
+                null_mut(),
+                &mut shared_secret_len,
+                ciphertext.as_mut_ptr(),
+                ciphertext.len(),
+            ) {
+                return Err(Unspecified);
+            }
+            let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+            if EVP_PKEY_decapsulate(
+                *ctx,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+                ciphertext.as_mut_ptr(),
+                ciphertext.len(),
+            ) != 1
+            {
+                shared_secret.zeroize();
+                return Err(Unspecified);
+            }
+            kdf(&shared_secret)
+        }
+    }
+
+    /// Creates a new KEM private key from raw bytes
+    ///
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPrivateKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    ///
+    pub fn new(alg: KemAlgorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
```

Comment:
We should make it clear in the documentation that `KemPrivateKey` should generally be created from `KemPrivateKey::generate`. This is only used to recreate private keys that had been previously generated and stored.